### PR TITLE
Botinator

### PR DESCRIPTION
**Pull request by MiladNazeri**
https://hifi-content.s3.amazonaws.com/Experiences/Releases/usefulUtilities/botinator/2019-03-30_22-37-48/botinator.js

### CODE REVIEW

**Comment by AndySeattle** [comment]
Feedback was as follows (and addressed):

- Audio cannot currently be adjusted for bot due to platform bug, so this slider needs to be removed. 
- Avatars can only be stopped and not removed by app. (So server restart needed to clear the board) This is due to a platform bug. 
- Application was spamming debug log while open and using create (addressed.)
- Location was random but bots would cluster. (addressed.) 

Please notify me if there is a major structural change in code review. I will re-test if needed, for now QA approved pending CR.

File: usefulUtilities/botinator/assignmentClientManager.js [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    // END UTILITY FUNCTIONS
+    // *************************************
+
+    // *************************************
+    // START CONSTS_AND_VARS
+    // *************************************
+    // #region CONSTS_AND_VARS
+
+
+    var X = 0;
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientManager.js]: ⚠This feels dangerous to me, especially because you're doing things like `x[0]` in your code. You're relying on case sensitivity, which works for computers (most of the time), but doesn't always work for humans reading the code. Can you make a case for needing these?

File: usefulUtilities/botinator/assignmentClientManager.js [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        this.position = position;
+        this.volume = volume;
+    }
+
+
+    // Sets the position to play this bot at
+    function setPosition(position) {
+        this.position = position;
+    }
+
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientManager.js]: insert one extra carriage return

File: usefulUtilities/botinator/assignmentClientManager.js [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+                botRegisterdCount++;
+                var messageToSend = JSON.stringify({
+                    action: "AC_AVAILABLE_UPDATE",
+                    newAvailableACs: availableAssignmentClientPlayers.length
+                });
+                Messages.sendMessage(ASSIGNMENT_CLIENT_MESSANGER_CHANNEL, messageToSend);
+                break;
+
+
+            case "ARE_YOU_THERE_MANAGER_ITS_ME_BOT":
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientManager.js]: 😂😂😂

File: usefulUtilities/botinator/assignmentClientManager.js [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+    // *************************************
+    // START MAIN
+    // *************************************
+    // #region MAIN
+    
+    
+    // Startup for the manager when it comes online
+    function startUp(){
+        Messages.subscribe(ASSIGNMENT_MANAGER_CHANNEL);
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientManager.js]: make sure you unsubscribe from both of these channels in `onEnding`

File: usefulUtilities/botinator/assignmentClientManager.js [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+    // *************************************
+
+    // *************************************
+    // START MAIN
+    // *************************************
+    // #region MAIN
+    
+    
+    // Startup for the manager when it comes online
+    function startUp(){
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientManager.js]:
```suggestion
    function startUp() {
```

please look for other instances of this fix in your files and modify them :)

File: usefulUtilities/botinator/assignmentClientManager.js (hunk 4 above)
Comment (zfox23) [reply usefulUtilities/botinator/assignmentClientManager.js]: (of course, after you send that final message)

File: usefulUtilities/botinator/assignmentClientPlayer.js [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            sender === scriptUUID || 
+            IGNORE_PLAYER_MESSAGES.indexOf(message.action) > -1) {
+            return;
+        }
+
+        try {
+            message = JSON.parse(message);
+        } catch (e) {
+            console.log("Can not parse message object");
+            console.log(e);
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientPlayer.js]: do you mean to return here?

File: usefulUtilities/botinator/assignmentClientPlayer.js [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        }
+
+        if (message.uuid !== scriptUUID) {
+            return;
+        }
+
+        switch (message.action){
+
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientPlayer.js]: unnecessary whitespace?

File: usefulUtilities/botinator/assignmentClientManager.js [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+            message = JSON.parse(message);
+        } catch (error) {
+            console.log("invalid object");
+            console.log("MESSAGE:", message);
+
+            return;
+        }
+
+        switch (message.action) {
+
```
Comment (zfox23) [inline usefulUtilities/botinator/assignmentClientManager.js]: unnecessary whitespace?

File: usefulUtilities/botinator/botinator.js [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+    
+    // *************************************
+    // START CONSTS_AND_VARS
+    // *************************************
+    // #region CONSTS_AND_VARS
+    
+
+    var ASSIGNMENT_CLIENT_MESSANGER_CHANNEL = "ASSIGNMENT_CLIENT_MESSANGER_CHANNEL";
+
+    var X = 0;
```
Comment (zfox23) [inline usefulUtilities/botinator/botinator.js]: same comment as before

File: usefulUtilities/botinator/tabletApp/js/botinator_ui.js [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+var loadingContainer;
+var contentBoundaryCorners1Position;
+var contentBoundaryCorners2Position;
+var volumeSlider;
+var totalNumberOfBotsNeeded;
+var availableACs;
+var playStopButton; 
+var updateBotDataButton;
+var playButtonDisabled = true;
```
Comment (zfox23) [inline usefulUtilities/botinator/tabletApp/js/botinator_ui.js]: any reason you can't replace instances of this with `playStopButton.disabled` ?

File: usefulUtilities/botinator/assignmentClientManager.js (hunk 1 above)
Comment (MiladNazeri) [reply usefulUtilities/botinator/assignmentClientManager.js]: It's me being lazy.  Lol there is no other way around it.  A part of me just likes the eaze of being able to do array operations on vectors, but I get it, it's an extra step and makes the code harder to read instead of just doing the normal x, y, z vector.  I'll update.

File: usefulUtilities/botinator/assignmentClientManager.js (hunk 3 above)
Comment (MiladNazeri) [reply usefulUtilities/botinator/assignmentClientManager.js]: That put a special smile on my face

File: usefulUtilities/botinator/assignmentClientManager.js (hunk 4 above)
Comment (MiladNazeri) [reply usefulUtilities/botinator/assignmentClientManager.js]: good call.  done.

File: usefulUtilities/botinator/assignmentClientPlayer.js (hunk 7 above)
Comment (MiladNazeri) [reply usefulUtilities/botinator/assignmentClientPlayer.js]: done

File: usefulUtilities/botinator/assignmentClientManager.js (hunk 5 above)
Comment (MiladNazeri) [reply usefulUtilities/botinator/assignmentClientManager.js]: donzo

File: usefulUtilities/botinator/tabletApp/js/botinator_ui.js (hunk 10 above)
Comment (MiladNazeri) [reply usefulUtilities/botinator/tabletApp/js/botinator_ui.js]: I completely forgot it's a getter/setter
